lib/sshutils/x11/forward.go

```diff
@@ -0,0 +1,229 @@
+	channels := clt.HandleChannelOpen(sshutils.X11ChannelRequest)
+	if channels == nil {
+		return trace.AlreadyExists("x11 forwarding channel already open")
+	}
+
+	go func() {
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+		for {
+			select {
+			case nch := <-channels:
+				if nch == nil {
+					return
+				}
+				go handler(ctx, nch)
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	return nil
+}
+
+// OpenNewXServerListener opens a new XServer unix socket with the first unused display
+// number. The unix socket's corresponding XServer display value will also be returned.
+func OpenNewXServerListener(displayOffset int, screen uint32) (net.Listener, string, error) {
+	for displayNumber := displayOffset; displayNumber < displayOffset+x11MaxDisplays; displayNumber++ {
+		if l, err := net.Listen("unix", xserverUnixSocket(displayNumber)); err == nil {
+			return l, unixXDisplay(displayNumber), nil
+		}
+	}
+
+	return nil, "", trace.LimitExceeded("No more x11 sockets are available")
+}
+
+// Forward begins x11 forwarding between an xserver connection and an x11 channel.
+// The xserver connection may be a direct xserver connection, or another x11 channel.
+func Forward(xconn io.ReadWriteCloser, x11Chan ssh.Channel) {
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		io.Copy(x11Chan, xconn)
+		x11Chan.CloseWrite()
+	}()
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		io.Copy(xconn, x11Chan)
+		// prefer CloseWrite over Close to prevent reading from halting pre-maturely.
+		switch c := xconn.(type) {
+		case interface{ CloseWrite() error }:
+			c.CloseWrite()
+		default:
+			c.Close()
+		}
+	}()
+	wg.Wait()
+}
+
+// DialXDisplay connects to the local xserver set in the $DISPLAY variable.
+func DialXDisplay(display string) (net.Conn, error) {
+	hostname, displayNumber, _, err := parseDisplay(display)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// If display is a unix socket, dial the address as an x11 unix socket
+	if hostname == "unix" || hostname == "" {
+		return net.Dial("unix", xserverUnixSocket(displayNumber))
+	}
+
+	// If hostname can be parsed as an IP address, dial the address as an x11 tcp socket
+	if ip := net.ParseIP(hostname); ip != nil {
+		conn, err := net.Dial("tcp", xserverTCPSocket(hostname, displayNumber))
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return conn, nil
+	}
+
+	// dial display as generic socket address
+	return net.Dial("unix", display)
+}
+
+// parsesDisplay parses the given display value and returns the host,
+// display number, and screen number, or a parsing error. display
+// should be in the format "hostname:displayNumber.screenNumber".
+func parseDisplay(display string) (string, int, int, error) {
+	splitHost := strings.Split(display, ":")
+	host := splitHost[0]
+	if len(splitHost) < 2 {
+		return host, 0, 0, nil
+	}
+
+	splitDisplayNumber := strings.Split(splitHost[1], ".")
+	displayNumber, err := strconv.Atoi(splitDisplayNumber[0])
```

Comment:
AFAIU this works with display numbers from 0-9. Any changes to go over that?

Reply:
If `display=unix:10.1`, then `splitDisplayNumber=["10","1"], so this should work. I'll change the naming to make it less confusing